esmvalcore/cmor/_fixes/common.py

```diff
@@ -108,49 +109,70 @@ def fix_metadata(self, cubes):
-        rlat_idx_bnds[len(rlat) - 1, 1] = len(rlat)
-        rlon_idx_bnds = np.zeros((len(rlon), 2))
-        rlon_idx_bnds[:, 0] = np.arange(len(rlon)) - 0.5
-        rlon_idx_bnds[:, 1] = np.arange(len(rlon)) + 0.5
+        cube = self.get_cube_from_list(cubes)
+
+        # Get dimensional coordinates. Note:
+        # - First dimension i -> X-direction (longitude; = dimension 2 in data)
+        # - Second dimension j -> Y-direction (latitude; = dimension 1 in data)
+        i_coord = cube.coord(dim_coords=True, dimensions=2)
+        j_coord = cube.coord(dim_coords=True, dimensions=1)
+
+        # Fix metadata of coordinate i
+        i_coord.var_name = 'i'
+        i_coord.standard_name = None
+        i_coord.long_name = 'cell index along first dimension'
+        i_coord.units = '1'
+        i_coord.circular = False
+
+        # Fix metadata of coordinate j
+        j_coord.var_name = 'j'
+        j_coord.standard_name = None
+        j_coord.long_name = 'cell index along second dimension'
+        j_coord.units = '1'
+
+        # Guess bounds of coordinates i and j if necessary
+        if not i_coord.has_bounds():
+            i_coord.guess_bounds()
+        if not j_coord.has_bounds():
+            j_coord.guess_bounds()
+
+        # Transform i and j bounds into array indices [0, 1, 2, ...]
+        i_to_idx = InterpolatedUnivariateSpline(i_coord.points,
+                                                np.arange(len(i_coord.points)),
+                                                k=1)
+        j_to_idx = InterpolatedUnivariateSpline(j_coord.points,
+                                                np.arange(len(j_coord.points)),
+                                                k=1)
+        i_idx_bnds = i_to_idx(i_coord.bounds)
+        j_idx_bnds = j_to_idx(j_coord.bounds)
 
         # Calculate latitude/longitude vertices by interpolation
         lat_vertices = []
         lon_vertices = []
-        for (i, j) in [(0, 0), (0, 1), (1, 1), (1, 0)]:
-            (rlat_v, rlon_v) = np.meshgrid(rlat_idx_bnds[:, i],
-                                           rlon_idx_bnds[:, j],
-                                           indexing='ij')
+        for (j, i) in [(0, 0), (0, 1), (1, 1), (1, 0)]:
+            (j_v, i_v) = np.meshgrid(j_idx_bnds[:, j],
+                                     i_idx_bnds[:, i],
+                                     indexing='ij')
             lat_vertices.append(
                 map_coordinates(cube.coord('latitude').points,
-                                [rlat_v, rlon_v],
+                                [j_v, i_v],
                                 mode='nearest'))
             lon_vertices.append(
                 map_coordinates(cube.coord('longitude').points,
-                                [rlat_v, rlon_v],
+                                [j_v, i_v],
                                 mode='wrap'))
         lat_vertices = np.array(lat_vertices)
         lon_vertices = np.array(lon_vertices)
```

Comment:
This box should also refer to the following two lines 179 and 180, but github wouldn't let me. Please take this into consideration.

So the assumption here is that we can construct reasonable bounds by placing corners exactly in-between the centers of the cells marked by the points. I am not completely sure that works because of overlapping and degenerate cells (for an example of this, have a close look at the two northernmost rows of longitudes in EC-Earth-3 Ofx areacello around the singularities), but it may be the best we can do.

In that case, I think it is easier (and avoids looping) to construct the entire field of corners and then sort it into the bounds like this:
```suggestion
# Assuming i->lon, j->lat !
corner_indices = np.meshgrid(i_idx_bnds, j_idx_bnds)
lon_corners = map_coordinates(cube.coord('longitude').points, corner_indices)
lat_corners = map_coordinates(cube.coord('latitude').points, corner_indices)
lon_vertices = np.stack([lon_corners[0:-1, 0:-1], lon_corners[0:-1, 1:], lon_corners[1:, 1:], lon_corners[1:, 0:-1]], axis=-1)
lat_vertices = np.stack([lat_corners[0:-1, 0:-1], lat_corners[0:-1, 1:], lat_corners[1:, 1:], lat_corners[1:, 0:-1]], axis=-1)

```

Reply:
> So the assumption here is that we can construct reasonable bounds by placing corners exactly in-between the centers of the cells marked by the points. I am not completely sure that works because of overlapping and degenerate cells (for an example of this, have a close look at the two northernmost rows of longitudes in EC-Earth-3 Ofx areacello around the singularities), but it may be the best we can do.

This is only true if the index coordinate do not have bounds. If they do, e.g., for the BCC models, this should not be an issue.

Regarding your suggestion: While I agree this is probably faster than the original version, I think it's less readable. This code is not run a thousand times and the `for` loop contains only 4 elements, so this really shouldn't be a performance issue. I'd like to leave it as is for the moment.

Reply:
I agree that readability is the overriding concern. That's precisely why I suggested the changes :smile: 

But it seems a prior we find different things easier to read. Perhaps that is an indication that some comment is in order, maybe with a link to the relevant CF section. Note that CF-1.9 will include [this picture](https://raw.githubusercontent.com/cf-convention/cf-conventions/40aef55dea53dc9402111dac975c0ef3b3e9bbbf/images/order_horizontal_bounds__2D_coord_variables.png) to explain the boundaries.

Reply:
Better link: http://cfconventions.org/cf-conventions/cf-conventions.html#cell-boundaries

Reply:
Done

Reply:
In terms of readability, note that in the loop-based solution (almost) every corner is calculated four times, but it is not obvious that the four independent calculations lead to the same values unless you are acutely aware of the values of bounds. That is the main thing that I think makes it harder to really understand while reading. I have changed my suggestion above to be a bit more readable, perhaps you want to reconsider.

In terms of the comments, I think it would be good to add a few inline, like `# going counter-clockwise around the cell` or maybe change the general comment to something like
```
Following the CF conventions (see ...), we go counter-clockwise around the cells and construct a grid of index values which are in turn used to interpolate longitudes and latitudes in the midpoints between the cell centers.
```

Reply:
I tried your solution. Unfortunately, the tests fails now:

```
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=0
E       
E       Mismatched elements: 20 / 24 (83.3%)
E       Max absolute difference: 40.
E       Max relative difference: 1.27272727
E        x: array([[[  0.  ,   0.  , -23.75,   0.  ],
E               [  0.  ,   0.  ,   0.  , -23.75],
E               [  0.  ,   0.  ,   0.  ,   0.  ]],...
E        y: array([[[-4.000000e+01, -3.375000e+01, -2.375000e+01, -3.000000e+01],
E               [-3.375000e+01, -6.250000e+00,  3.750000e+00, -2.375000e+01],
E               [-6.250000e+00, -1.024181e-14,  1.000000e+01,  3.750000e+00]],...
```

To be honest I really would like to keep to original solution. I don't think there's a risk that the four calculations give different results since we know exactly how the 1D bounds look like. I also still think that the original solution is more readable.

I'm happy to extend the comments, but I suggest that any further changes to this function should be implemented in a separate PR.

Reply:
Well, of course, it's just the illustration of an idea typed into a web browser, so I am more surprised that it runs, than that it doesn't produce the expected output.

I also didn't say that the calculations produce different numbers. What I meant is that you need to understand and keep in mind a lot more lines of code to convince yourself of that fact.

But if you prefer, keep the loop-based version. I think the comments you put in place help :+1:.

I am slightly concerned about the boundary handling, which probably means that the northernmost and southernmost cells are likely cut in half, possibly conflicting with the information in areacello. But let's leave that for another PR as well.